images/dnsmasq/validate_dynamic_deps.py

```diff
@@ -0,0 +1,170 @@
+#/usr/bin/env python3
+
+# Copyright 2022 The Kubernetes Authors.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+# http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# This script requires python 3.8+
+import os
+import shutil
+import subprocess
+import pathlib
+import sys
+from typing import Optional
+
+IMAGE_DIR = '.test-images'
+
+
+def replace_img_special_chars(image_ref: str) -> str:
+    """Replace container image name characters that cannot be a part of container names or file names."""
+    return image_ref.lower().replace('/', '_').replace(':', '_')
+
+
+def relative_to_root(path: pathlib.Path) -> pathlib.Path:
+    return path.relative_to(path.anchor)
+
+
+def detect_architecture(cwd: pathlib.Path, target_binary: pathlib.Path) -> str:
+    """Read the architecture of the ELF file target_binary using the file command"""
+    result = subprocess.run(['file', target_binary],
+                            capture_output=True,
+                            text=True,
+                            cwd=cwd)
+    return result.stdout.split(",")[1].strip()
+
+
+def untar_container_image(image_ref: str, output_dir: pathlib.Path):
+    """Create a docker container using image_ref use tar to extract the filesystem to output_dir."""
+    container_name = replace_img_special_chars(image_ref)
+    subprocess.run(['docker', 'create', '--name', container_name, image_ref],
+                   capture_output=True)
+    subprocess.run(f'docker export {container_name} | tar x',
+                   shell=True,
+                   cwd=output_dir,
+                   capture_output=True)
+    subprocess.run(['docker', 'rm', container_name], capture_output=True)
+
+
+def detect_dependencies(cwd: pathlib.Path,
+                        target_binary: pathlib.Path) -> [str]:
+    """Use objdump to read the ELF headers for target_binary and return a list of .so files."""
+    result = subprocess.run(['objdump', '--private-headers', target_binary],
+                            capture_output=True,
+                            text=True,
+                            cwd=cwd)
+    deps = []
+    for line in result.stdout.splitlines():
+        if "NEEDED" in line:
+            deps.append(line.split()[1])
+    return deps
+
+
+def find_dependency_path(root_dir: pathlib.Path, dependency_name: str)-> Optional[pathlib.Path]:
+    """Search for the first file called dependency_name within root_dir."""
+    paths = sorted(root_dir.glob(f'**/{dependency_name}'))
```

Comment:
Shared object files are resolved by name, so we should only need the first file out of the list. This will not work if the container had multiple copies of the required `.so` files.

While I was looking into less naive ways of resolving these dependencies ,I noticed that the ldconfig tool has `-r` option to use the container's fs as a root. We can use it to load all of the .so files in the image. I'll update the script to use that instead. it should be more reliable.

Reply:
`ldconfig` worked on multiple architectures, so I used it to replace this function. Let me know what you think of the new dependency resolution strategy.

Reply:
Nevermind, ldconfig can read cache files generated on other architectures, but it doesn't when generating new cache files.

I'll revert back to the globbing which I was doing before and make it handle multiple `.so` files with the same name.
```
❯ ldconfig -r .test-images/gcr.io_k8s-staging-dns_k8s-dns-dnsmasq-arm64_1.22.5-dirty -v   
ldconfig: Can't stat /lib/x86_64-linux-gnu: No such file or directory
ldconfig: Can't stat /usr/lib/x86_64-linux-gnu: No such file or directory
/lib: (from <builtin>:0)
ldconfig: /lib/ld-linux-aarch64.so.1 is for unknown machine 183.

/usr/lib: (from <builtin>:0)
```